File: sach/de3398.py
//DE3398.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
from bisect import bisect_left, bisect_right
N = 200001
n, q = map(int, input().split())
id = [0] * N
save = [[] for _ in range(N)]
for i in range(1, n + 1):
    x = int(input())
    id[x] = i
for i in range(1, N):
    for j in range(i, N, i):
        if id[j]:
            save[i].append(id[j])
        if id[i]:
            save[j].append(id[i])
for i in range(1, N):
    save[i].sort()
    save[i] = list(dict.fromkeys(save[i]))
for _ in range(q):
    l, r, d = map(int, input().split())
    u = bisect_left(save[d], l)
    v = bisect_right(save[d], r)
    print(v - u, end=" ")
print()

File: sach/de3398.cpp
//DE3398.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
#define all(x) x.begin(), x.end()
typedef long long ll;
const int N = 2e5 + 1;
int n, q, id[N];
vector <int> save[N];
int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    cin >> n >> q;
    for(int x, i = 1; i <= n; i++) {
        cin >> x;
        id[x] = i;
    }
    for(int i = 1; i < N; i++) {
        for(int j = i; j < N; j += i) {
            if (id[j]) save[i].push_back(id[j]);
            if (id[i]) save[j].push_back(id[i]);
        }
    }
    for(int i = 1; i < N; i++) {
        sort(all(save[i]));
        save[i].erase(unique(all(save[i])), save[i].end());
    }
    for(int i = 1; i <= q; i++) {
        int l, r, d; cin >> l >> r >> d;
        int u = lower_bound(all(save[d]), l)-save[d].begin();
        int v = upper_bound(all(save[d]), r)-save[d].begin();
        cout << v - u << " ";
    }
    cout << "\n";
}

